### ISSUE TITLE
Make Edit Objects more useful

### ISSUE THREAD

**Issue by eevee** [sirjuddington/SLADE#331]
"Edit Objects" is great, but it's also hard to use with a selection that's connected to unselected lines, since those lines also move or rotate.  If I want to move a doorway from the right side of a room to the bottom, for example, I have to do a lot of extra work inserting extra vertices and making sure there's enough room to rotate without getting overlapping walls.  It'd be really helpful if "Edit Objects" had a key for _detaching_ a selection from its surroundings as unobtrusively as possible.

This would also help for small edits that are intuitive to humans but not to the Doom map format, like moving a crate out of the corner of a room.  _I_ think of this as moving a crate, but Doom thinks of it as seriously rearchitecting two independent sectors:

![slade-crate-corner](https://cloud.githubusercontent.com/assets/94112/10311637/6ec303f8-6bfc-11e5-9d39-aabf530c9c22.png) ![slade-crate-middle](https://cloud.githubusercontent.com/assets/94112/10311638/6ec3f1a0-6bfc-11e5-98d8-966c363d590a.png)

Similarly, it would be helpful if I could paste a selection directly into Edit Objects mode, so I could e.g. make rotated versions of a feature without having to separately paste into a temporary empty area.

**Comment by sirjuddington** [comment]
It's probably possible to add the ability to 'detach' the objects being edited, however I'm not sure that example you posted would be, at least without a considerable amount of 'guessing' on SLADE's part (which would probably end up wrong half the time). Wouldn't be too hard to have it work like that but leave a 'gap' in those outer lines where the crate was.

**Comment by eevee** [comment]
Yeah, it's definitely something that would be heuristic at best.  I _think_ the human-friendly approach would be to check whether the selection is more inside or outside of its neighboring sectors.  Consider:

![slade-crate-corner1](https://cloud.githubusercontent.com/assets/94112/10384791/bed3f1a4-6df5-11e5-99ac-22f650bf98f8.png)

To my eye, the top crate is entirely "inside" the room and the bottom crate is entirely "outside", because I see all those vertical lines as a single unbroken wall.  So you'd have to compute the polygon that _would_ exist if the crates weren't there, by extending its other lines to fill the gaps, and then see that the top crate is entirely inside that hypothetical sector but the bottom one is entirely outside of it.  Should do what you mean for common structures like inset lighting, recessed switches, COMP junk, and the like.

That also extends fairly well to cases where there's more than one neighbor, for example a doorway:

![slade-crate-corner1](https://cloud.githubusercontent.com/assets/94112/10384900/15821f16-6df7-11e5-879b-f02539d3065f.png)

That middle sector is entirely outside both rooms, so detaching it should function like deleting it, leaving the two rooms unconnected.

Then there are more, uh, exotic cases:

![slade-crate-corner1](https://cloud.githubusercontent.com/assets/94112/10384971/c5863118-6df7-11e5-96e2-844731de91aa.png)

I'm not sure what the right thing is here!  Delete them all and leave a void?  But they're all part of the rooms, by the above definition.  Should detaching them fill in the space and leave four triangular rooms?  What if the virtual rooms overlapped awkwardly?

Interesting problem, at least  :)

**Comment by Gaerzi** [comment]
The only sane approach, if you don't want to recruit an entire team of computer scientists, would be to keep the other sectors exactly like they are, as if you had cut the detached sector entirely, and then pasted it elsewhere.

This mean that, in your first example, the parent sector would keep a hole in its southwest corner.

**Comment by eevee** [comment]
Another major use case for this: the "flip" operations will generally produce completely garbage geometry when used on connected areas.  It'd be great if they automatically detached the selection first, but being able to do it manually would at least be a stopgap.
